Dockerfile 추가

base image를 `maven:3-alpine` 으로 docker build 작업 수행하면 dependency download가 일부 누락되서
`mvn -o package` 수행이 안되는 문제 있었습니다.
같은 내용을 `maven:3.5-jdk-8` 로 수행하면 해결됩니다. 이유는 모르겠습니다.

### Dockerfile 내용
```dockerfile
FROM maven:3-alpine as config-service-build

WORKDIR /app

ADD pom.xml pom.xml
RUN mvn clean dependency:go-offline

ADD src src
RUN mvn -o package


FROM openjdk:8u171-jre-alpine
WORKDIR /app

COPY --from=config-service-build /app/target/*.jar .

EXPOSE 8088

CMD java -jar *.jar
```

### 에러메세지
```bash
$ docker build \
    --target config-service-build \
    -t config-service-build \
    .

...
Step 6/6 : RUN mvn -o package
 ---> Running in 81fe5fe6d62f
[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for pl.piomin.services:config-service:1.0-SNAPSHOT: Cannot access central (https://repo.maven.apache.org/maven2) in offline mode and the artifact org.springframework.boot:spring-boot-starter-parent:pom:2.0.6.RELEASE has not been downloaded from it before. and 'parent.relativePath' points at wrong local POM @ line 3, column 10
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project pl.piomin.services:config-service:1.0-SNAPSHOT (/app/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for pl.piomin.services:config-service:1.0-SNAPSHOT: Cannot access central (https://repo.maven.apache.org/maven2) in offline mode and the artifact org.springframework.boot:spring-boot-starter-parent:pom:2.0.6.RELEASE has not been downloaded from it before. and 'parent.relativePath' points at wrong local POM @ line 3, column 10 -> [Help 2]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
The command '/bin/sh -c mvn -o package' returned a non-zero code: 1
```

`dive`라는 바이너리가 있는데요, 그걸로 도커 이미지를 검사 할 수 있습니다.
![image](https://user-images.githubusercontent.com/33619494/56556433-fc2f3300-65d2-11e9-8b10-6efb3c1643d3.png)

이건 `maven:3-alpine` 검사한거구요.

![image](https://user-images.githubusercontent.com/33619494/56556488-1a952e80-65d3-11e9-990a-8398e6cb9372.png)
이건 `maven:3.5-jdk-8` 검사한겁니다.

둘 차이점이 os도 있지만 일단 maven 버전이 다르네요.

근데 저는 `maven:3-alpine`이미지에서 `mvn -o package` 명령 쳐도 빌드가 잘 되는데요
![image](https://user-images.githubusercontent.com/33619494/56557639-ca1fd000-65d6-11e9-826d-3be2e518233a.png)

`Dockerfile`로 만드니 에러가 발생하네요..
```
[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for pl.piomin.services:config-service:1.0-SNAPSHOT: Cannot access central (https://repo.maven.apache.org/maven2) in offline mode and the artifact org.springframework.boot:spring-boot-starter-parent:pom:2.0.6.RELEASE has not been downloaded from it before. and 'parent.relativePath' points at wrong local POM @ line 3, column 10
 @
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]
[ERROR]   The project pl.piomin.services:config-service:1.0-SNAPSHOT (/app/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for pl.piomin.services:config-service:1.0-SNAPSHOT: Cannot access central (https://repo.maven.apache.org/maven2) in offline mode and the artifact org.springframework.boot:spring-boot-starter-parent:pom:2.0.6.RELEASE has not been downloaded from it before. and 'parent.relativePath' points at wrong local POM @ line 3, column 10 -> [Help 2]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
The command '/bin/sh -c mvn -o package' returned a non-zero code: 1
```

이런 귀찮음을 방지하기 위해 그냥 저는 latest 태그 달린 것 씁니다..

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,8 +1,19 @@
-FROM openjdk:8-jre-alpine
-ENV APP_FILE config-service-1.0-SNAPSHOT.jar
-ENV APP_HOME /app
+FROM maven:3.5-jdk-8 as config-service-build
+
+WORKDIR /app
+
+ADD pom.xml pom.xml
+RUN mvn clean dependency:go-offline
+
+ADD src src
+RUN mvn -o package
+
+
+FROM openjdk:8u171-jre-alpine
+WORKDIR /app
+
+COPY --from=config-service-build /app/target/*.jar .
+
 EXPOSE 8088
-COPY target/$APP_FILE $APP_HOME/
-WORKDIR $APP_HOME
-ENTRYPOINT ["sh", "-c"]
-CMD ["exec java -jar $APP_FILE"]
+
+CMD java -jar *.jar

--- a/tools/docker/build_docker.sh
+++ b/tools/docker/build_docker.sh
@@ -1,0 +1,50 @@
+#!/bin/bash -e
+
+yellow="\033[0;33m"
+red="\033[0;31m"
+nocolor="\033[0m"
+
+command_exists() {
+    command -v "$@" > /dev/null 2>&1
+}
+
+get_script_path() {
+    local _src="${BASH_SOURCE[0]}"
+    while [[ -h "${_src}" ]]; do
+        local _dir="$(cd -P "$( dirname "${_src}" )" && pwd)"
+        local _src="$(readlink "${_src}")"
+        if [[ "${_src}" != /* ]]; then _src="$_dir/$_src"; fi
+    done
+    echo $(cd -P "$(dirname "$_src")" && pwd)
+}
+
+cd_into_script_path() {
+    local script_path=$(get_script_path)
+    cd ${script_path}
+}
+
+
+### Main
+echo -e "${yellow}[INFO] [${BASH_SOURCE[0]}] executed.${nocolor}"
+
+if ! command_exists docker; then
+    echo -e "${red}[ERROR] Install docker first.\n\
+[MacOS] $ brew cask install docker\n\
+[Linux] $ curl -fsSL https://get.docker.com | sh${nocolor}"
+    exit 1
+fi
+
+(
+cd_into_script_path
+cd ../..
+
+docker build \
+    --target config-service-build \
+    -t config-service-build \
+    .
+
+docker build \
+    -t config-service \
+    .
+)
+echo -e "${yellow}[INFO] [${BASH_SOURCE[0]}] Done.${nocolor}"

--- a/tools/docker/push_docker.sh
+++ b/tools/docker/push_docker.sh
@@ -1,0 +1,69 @@
+#!/bin/bash -e
+
+yellow="\033[0;33m"
+red="\033[0;31m"
+nocolor="\033[0m"
+
+command_exists() {
+    command -v "$@" > /dev/null 2>&1
+}
+
+get_script_path() {
+    local _src="${BASH_SOURCE[0]}"
+    while [[ -h "${_src}" ]]; do
+        local _dir="$(cd -P "$( dirname "${_src}" )" && pwd)"
+        local _src="$(readlink "${_src}")"
+        if [[ "${_src}" != /* ]]; then _src="$_dir/$_src"; fi
+    done
+    echo $(cd -P "$(dirname "$_src")" && pwd)
+}
+
+cd_into_script_path() {
+    local script_path=$(get_script_path)
+    cd ${script_path}
+}
+
+get_version() {
+    echo $(mvn -q \
+        -Dexec.executable=echo \
+        -Dexec.args='${project.version}' \
+        --non-recursive \
+        exec:exec)
+}
+
+### Main
+echo -e "${yellow}[INFO] [${BASH_SOURCE[0]}] executed.${nocolor}"
+
+if ! command_exists docker; then
+    echo -e "${red}[ERROR] Install docker first.\n\
+[MacOS] $ brew cask install docker\n\
+[Linux] $ curl -fsSL https://get.docker.com | sh${nocolor}"
+    exit 1
+fi
+
+if ! command_exists mvn; then
+    echo -e "${red}[ERROR] install maven first. \n\
+[Mac] brew install maven \n\
+[Centos] yum install -y maven \n\
+[Ubuntu] apt-get install -y maven${nocolor}"
+    exit 1;
+fi
+
+(
+cd_into_script_path
+./build_docker.sh
+)
+
+(
+cd_into_script_path
+cd ../..
+
+echo -e "${yellow}[INFO] getting version...${nocolor}"
+version=$(get_version)
+echo -e "${yellow}[INFO] getting version done. [verison=$version]${nocolor}"
+
+docker tag config-service kujyp/config-service:$version
+docker push kujyp/config-service:$version
+)
+
+echo -e "${yellow}[INFO] [${BASH_SOURCE[0]}] Done.${nocolor}"

